dataset.go

```diff
@@ -86,6 +114,70 @@ func flatElementsIterator(elems []*Element, elemChan chan<- *Element) {
 	}
 }
 
+// FlatDatasetIterator represents a simple iterator over the elements in a
+// Dataset.
+type FlatDatasetIterator interface {
+	// HasNext indicates if the iterator as another element.
+	HasNext() bool
+	// Next gets and returns the next element in the iterator.
+	Next() *Element
+}
+
+// FlatStatefulIterator returns a stateful iterator that adheres to
+// FlatDatasetIterator interface. This allows the caller to iterate over every
+// element in the dataset, including elements nested inside sequences.
+//
+// Important note: if the Dataset changes during the iteration (e.g. if elements
+// are added or removed), those elements will not be included until a new
+// iterator is created.
+//
+// If you don't need to receive elements on a/ channel, and don't want to worry
+// about always exhausting this iterator, this is the best and safest way to
+// iterate over a Dataset.
+func (d *Dataset) FlatStatefulIterator() FlatDatasetIterator {
+	flatElems := flatElemsWrapper{}
+	flatSliceBuilder(d.Elements, &flatElems)
+	// fmt.Println("Dataset:", flatElems.flatElems)
+	iter := flatIterator{flattenedDataset: flatElems.flatElems}
+	return &iter
+}
+
+func flatSliceBuilder(datasetElems []*Element, flatElems *flatElemsWrapper) {
+	for _, elem := range datasetElems {
+		if elem.Value.ValueType() == Sequences {
+			flatElems.append(elem)
+			for _, seqItem := range elem.Value.GetValue().([]*SequenceItemValue) {
+				flatSliceBuilder(seqItem.elements, flatElems)
+			}
+			continue
+		}
+		flatElems.append(elem)
+	}
+}
+
+type flatIterator struct {
+	flattenedDataset []*Element
+	idx              int
+}
+
+type flatElemsWrapper struct {
+	flatElems []*Element
+}
+
+func (f *flatElemsWrapper) append(e *Element) {
+	f.flatElems = append(f.flatElems, e)
+}
+
+func (f *flatIterator) HasNext() bool {
```

Comment:
If you returned a non-pointer this method could just take in `flatIterator` instead of `*flatIterator` and avoid a pointer dereference here.

Reply:
Similar to my response above, but typically we use pointer receivers on methods that modify the struct. In this case `Next()` does that. So, the rest of the methods should also be defined on the pointer as well.

More thoughts on this at https://github.com/golang/go/wiki/CodeReviewComments#receiver-type